File: projekat/backend/app/schemas/escalation.py
from __future__ import annotations

from datetime import datetime
from typing import List, Optional

from pydantic import BaseModel, ConfigDict


class EscalationInfo(BaseModel):
    escalation_id: int
    status: str
    queue_position: int
    trigger_tip: str


class EscalationRead(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    sesija_id: int
    korisnik_id: str
    dodjeljeni_agent_id: Optional[str] = None
    prioritet: str
    status: str
    trigger_tip: Optional[str] = None
    tema: Optional[str] = None
    razgovor: Optional[List[dict]] = None
    datum_kreiranja: datetime
    datum_rjesavanja: Optional[datetime] = None
    napomena_rjesavanja: Optional[str] = None
    queue_position: Optional[int] = None


class EscalationResolve(BaseModel):
    napomena: str = ""
    submit_to_kb: bool = False
    odgovor_agenta: Optional[str] = None
    pitanje_korisnika: Optional[str] = None


class AgentStatusUpdate(BaseModel):
    status: str

File: projekat/frontend/src/api/escalation.ts
import apiClient from "./client";
import type { AgentStats } from "../types";

export interface EscalationItem {
  id: number;
  sesija_id: number;
  korisnik_id: string;
  dodjeljeni_agent_id?: string | null;
  agent_name?: string | null;
  prioritet: string;
  status: string;
  trigger_tip?: string | null;
  tema?: string | null;
  razgovor: Array<{ role: string; content: string }>;
  datum_kreiranja: string;
  datum_rjesavanja?: string | null;
  napomena_rjesavanja?: string | null;
  queue_position?: number | null;
}

export interface EscalationResolvePayload {
  napomena: string;
  submit_to_kb: boolean;
  odgovor_agenta?: string;
  pitanje_korisnika?: string;
}

import type { EscalationInfo } from "../types";

export const escalationApi = {
  getQueue: () => apiClient.get<EscalationItem[]>("/api/v1/escalation/"),

  accept: (id: number) =>
    apiClient.post<{ ok: boolean; escalation: EscalationItem }>(
      `/api/v1/escalation/${id}/accept`
    ),

  resolve: (id: number, payload: EscalationResolvePayload) =>
    apiClient.post<{ ok: boolean }>(`/api/v1/escalation/${id}/resolve`, payload),

  release: (id: number) =>
    apiClient.post<{ ok: boolean }>(`/api/v1/escalation/${id}/release`),

  updateAgentStatus: (status: "Online" | "Zauzet" | "Offline") =>
    apiClient.patch("/api/v1/escalation/agent-status", { status }),

  request: (session_id: number, conversation_history: Array<{ role: string; content: string }>) =>
    apiClient.post<{ escalation: EscalationInfo }>("/api/v1/escalation/request", {
      session_id,
      conversation_history,
    }),

  cancel: () => apiClient.post<{ ok: boolean }>("/api/v1/escalation/cancel"),

  myStats: () => apiClient.get<AgentStats>("/api/v1/escalation/my-stats"),

  myHistory: (limit = 20, offset = 0) =>
    apiClient.get<EscalationItem[]>("/api/v1/escalation/my-history", {
      params: { limit, offset },
    }),
};
